File: job-matcher/src/scrapers/greenhouse-scraper.js
const BaseScraper = require('./base-scraper');

class GreenhouseScraper extends BaseScraper {
  async scrape() {
    const jobs = [];
    const selectors = this.boardConfig.selectors;
    const maxJobs = this.globalConfig.max_jobs_per_board || 50;
    
    try {
      console.log(`\nScraping Greenhouse board: ${this.boardConfig.name}`);
      await this.page.goto(this.boardConfig.url, { waitUntil: 'networkidle2' });
      
      await this.waitForSelector(selectors.job_list_container, 10000);
      await this.delay(2000);
      
      let hasNextPage = true;
      let pageNum = 1;
      
      while (hasNextPage && (maxJobs === 0 || jobs.length < maxJobs)) {
        console.log(`  Processing page ${pageNum}...`);
        
        const pageJobs = await this.page.evaluate((sel) => {
          const jobCards = document.querySelectorAll(sel.job_card);
          const extracted = [];
          
          jobCards.forEach(card => {
            const titleEl = card.querySelector(sel.title);
            const locationEl = card.querySelector(sel.location);
            
            if (titleEl) {
              const linkEl = titleEl.tagName === 'A' ? titleEl : card.querySelector('a');
              extracted.push({
                title: titleEl.textContent.trim(),
                location: locationEl ? locationEl.textContent.trim() : 'Not specified',
                url: linkEl ? linkEl.href : '',
                source: 'Greenhouse'
              });
            }
          });
          
          return extracted;
        }, selectors);
        
        // Get details for each job
        for (const job of pageJobs) {
          if (maxJobs > 0 && jobs.length >= maxJobs) break;
          
          try {
            const jobDetail = await this.getJobDetails(job);
            if (jobDetail) {
              jobs.push(jobDetail);
              console.log(`    ✓ ${job.title}`);
            }
          } catch (e) {
            console.log(`    ✗ Error getting details for ${job.title}: ${e.message}`);
          }
          
          await this.delay(1000);
        }
        
        // Try to go to next page
        hasNextPage = await this.goToNextPage(selectors.next_button);
        if (hasNextPage) {
          pageNum++;
          await this.delay(this.globalConfig.request_delay_ms || 2000);
        }
      }
      
    } catch (error) {
      console.error(`Error scraping Greenhouse board: ${error.message}`);
    }
    
    console.log(`  Total jobs scraped: ${jobs.length}`);
    return jobs;
  }
  
  async getJobDetails(job) {
    try {
      // Navigate to job detail page
      await this.page.goto(job.url, { waitUntil: 'networkidle2' });
      await this.delay(2000);
      
      const selectors = this.boardConfig.selectors;
      
      // Extract description
      const description = await this.page.evaluate(() => {
        // Try multiple selectors for Greenhouse job descriptions
        const selectors = [
          '#content',
          '.job-description',
          '[data-qa="job-description"]',
          '.description',
          '#job_description'
        ];
        
        for (const sel of selectors) {
          const el = document.querySelector(sel);
          if (el) {
            return el.textContent.trim();
          }
        }
        
        // Fallback: get all text content
        return document.body.innerText.substring(0, 3000);
      });
      
      return {
        ...job,
        description: description.substring(0, 3000),
        company: this.boardConfig.name
      };
    } catch (e) {
      return { ...job, description: '', company: this.boardConfig.name };
    }
  }
  
  async goToNextPage(nextButtonSelector) {
    if (!nextButtonSelector) return false;
    
    try {
      const nextButton = await this.page.$(nextButtonSelector);
      if (!nextButton) return false;
      
      const href = await nextButton.evaluate(el => el.getAttribute('href'));
      if (!href || href === '#' || href === '') return false;
      
      await nextButton.click();
      await this.delay(3000);
      return true;
    } catch (e) {
      return false;
    }
  }
}

module.exports = GreenhouseScraper;

File: job-matcher/src/scrapers/greenhouse_scraper.py
import json
import logging
import re
from urllib.parse import urljoin

log = logging.getLogger(__name__)


class GreenhouseScraper:
    def __init__(self, board_config):
        self.board_config = board_config

    async def scrape(self, base_scraper):
        url = self.board_config["url"]
        max_pages = self.board_config.get("max_pages", 10)
        known_urls = set()

        log.info("Scraping Greenhouse board: %s", self.board_config["name"])

        page = await base_scraper.new_page()
        try:
            await base_scraper.goto_with_retry(page, url)

            for page_num in range(1, max_pages + 1):
                log.info("  Processing page %d...", page_num)

                await page.wait_for_selector("section.posting-list", timeout=15000)
                job_cards = await page.query_selector_all(
                    "section.posting-list div.posting"
                )

                for card in job_cards:
                    link_el = await card.query_selector("a[href]")
                    if not link_el:
                        continue

                    href = await link_el.get_attribute("href")
                    if not href:
                        continue
                    job_url = urljoin(url, href)
                    if job_url in known_urls:
                        continue
                    known_urls.add(job_url)

                    title_el = await card.query_selector("h2, .title")
                    title = (await title_el.inner_text()).strip() if title_el else ""

                    loc_el = await card.query_selector(".location, .commitment")
                    location = (await loc_el.inner_text()).strip() if loc_el else ""

                    job_desc = ""
                    posted_date = ""
                    try:
                        detail_page = await base_scraper.new_page()
                        try:
                            await base_scraper.goto_with_retry(detail_page, job_url)
                            desc_el = await detail_page.query_selector("#content, .content, .job-description")
                            if desc_el:
                                job_desc = (await desc_el.inner_text()).strip()

                            body_text = await detail_page.inner_text("body")
                            m = re.search(r"Posted\s+(\d+)\s+(day|days|week|weeks)\s+ago", body_text, re.IGNORECASE)
                            if m:
                                posted_date = m.group(0)
                        finally:
                            await detail_page.close()
                    except Exception:
                        log.warning("  Could not load detail for %s", title)

                    yield {
                        "url": job_url,
                        "title": title,
                        "company": self.board_config.get("name", ""),
                        "location": location or "Not specified",
                        "description": job_desc or "Not available",
                        "source": "Greenhouse",
                        "postedDate": posted_date,
                    }

                next_btn = await page.query_selector(
                    "a[rel='next'], button:has-text('Next'), .pagination a:has-text('Next')"
                )
                if not next_btn:
                    log.info("  No more pages (no next button)")
                    break
                await next_btn.click()
                await page.wait_for_timeout(3000)

        finally:
            await page.close()
